src/txn/txn_ckpt.c

```diff
@@ -1439,12 +1440,29 @@ __checkpoint_lock_dirty_tree(
         }
     }
 
+    /*
+     * Discard the saved list of checkpoints, and slow path if this is not a WiredTiger checkpoint
+     * or if checkpoint drops are involved. Also, if we do not have checkpoint array size, the
+     * regular checkpoint process did not create the array, it is safer to discard the array in such
+     * a case.
+     */
+    if (!is_wt_ckpt || is_drop || btree->ckpt_allocated == 0)
+        __wt_meta_saved_ckptlist_free(session);
+
     /* If we have to process this btree for any reason, reset the timer and obsolete pages flag. */
     WT_BTREE_CLEAN_CKPT(session, btree, 0);
     F_CLR(btree, WT_BTREE_OBSOLETE_PAGES);
 
-    /* Get the list of checkpoints for this file. */
-    WT_ERR(__wt_meta_ckptlist_get(session, dhandle->name, true, &ckptbase));
+    /*
+     * Get the list of checkpoints for this file: We try to cache the ckptlist between the
+     * checkpoints. But there might not be one, as there are operations that can invalidate a
+     * ckptlist. So, use a cached ckptlist if there is one, otherwise go through slow path of
+     * re-generating the ckptlist by reading the metadata.
+     */
+    if (WT_IS_METADATA(dhandle))
+        WT_ERR(__wt_meta_ckptlist_get(session, dhandle->name, true, &ckptbase, &ckpt_allocated));
+    else if (__wt_meta_saved_ckptlist_get(session, dhandle->name, &ckptbase) != 0)
+        WT_ERR(__wt_meta_ckptlist_get(session, dhandle->name, true, &ckptbase, &ckpt_allocated));
```

Comment:
These could be combined as:
if (WT_IS_METADATA(dhandle) || __wt_meta_saved_ckptlist_get(...) != 0)
    WT_ERR(__wt_meta_ckptlist_get(...));

Reply:
good point. done.